File: src/VL.FSharpDemoLib/01_StaticOperations.fs
namespace FSharpDemoLib
open System.Runtime.InteropServices //for the parameter attributes

//use enum types not union types to be compatible with VL enums
type DemoEnum = Foo=0 | Bar=1

module SomeStaticVLNodes =

    //simple static operation node
    let MyAddition input input2 : float32 = input + input2

    //pinnames are separated at camelCasing for better reading in VL
    let PinNames firstInput secondInput : float32 = firstInput + secondInput

    //setting pin defaults, needs F# compiler tools 4.1 or Visual Studio 2017
    let Defaults ([<Optional; DefaultParameterValue(44)>] firstInput:float32) ([<Optional; DefaultParameterValue(0.44f)>] secondInput:float32) = firstInput + secondInput

    //multiple outputs via out parameters
    let MultipleOutputs (firstInput:float32) (secondInput:float32) ([<Out>] added:byref<float32>) ([<Out>] multiplied:byref<float32>) =
        added <- firstInput + secondInput
        multiplied <- firstInput * secondInput

    //using generics
    let Generic input = input.ToString()

    //seq<> appears as Sequence<> in vl
    let ReverseSequence input : seq<float32> = Seq.rev input

    //static enum
    let StaticEnumDemo (e:DemoEnum) = e.ToString()

    //to use XML documentation don't forget to enable "XML Documentation File" in the projects properties!
    ///<summary>An operation to test xml documentation</summary>
    ///<remarks>This is only an example node</remarks>
    ///<tags>html documentation test</tags>
    ///<param name="a">the param a</param>
    ///<returns>returns 2 a</returns>
    let XMLDocuTest a = a*2

    //you can use ref parameters, but
    //beware: assigning the parameter in the method leads to undefined behavior in VL (for now)
    //so only make use of it in a readonly fashion
    let RefParams (input:byref<int>) = input + 4444

File: src/VL.DemoLib/01_StaticOperations.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DemoLib
{
    //The name of a static class will show up as a category in VL
    public static class SomeStaticVLNodes
    { 
        //a simple static operation
        public static float MyAddition(float input, float input2)
        {
            return input + input2;
        }

        //function overloading
        public static float MyAddition(float input, float input2, float input3)
        {
            return input + input2 + input3;
        }

        //pinnames are separated at camelCasing for better reading in VL
        public static float PinNames(float firstInput, float secondInput)
        {
            return firstInput + secondInput;
        }

        //setting pin defaults
        public static float Defaults(float firstInput = 44f, float secondInput = 0.44f)
        {
            return firstInput + secondInput;
        }

        //multiple outputs via out parameters
        public static void MultipleOutputs(float firstInput, float secondInput, out float added, out float multiplied)
        {
            added = firstInput + secondInput;
            multiplied = firstInput * secondInput;
        }

        //using generics
        public static string Generic<T>(T input)
        {
            return input.ToString();
        }

        //IEnumerable<> appears as Sequence<> in vl
        public static IEnumerable<float> ReverseSequence(IEnumerable<float> input)
        {
            return input.Reverse();
        }

        //static enum
        public enum DemoEnum { Foo, Bar };
        public static string StaticEnumDemo(DemoEnum e)
        {
            return e.ToString();
        }

        //to use XML documentation don't forget to enable "XML Documentation File" in the projects properties!
        ///<summary>An operation to test xml documentation</summary>
        ///<remarks>This is only an example node</remarks>
        ///<tags>html documentation test</tags>
        ///<param name="a">the param a</param>
        ///<returns>returns 2 a</returns>
        public static int XMLDocuTest(int a)
        {
            return a*2;
        }

        //you can use ref parameters, but
        //beware: assigning the parameter in the method leads to undefined behavior in VL (for now)
        //so only make use of it in a readonly fashion
        public static int RefParams(ref int input)
        {
            return input + 4444;
        }
    }
}
